File: examples/messenger/perl/send_async.pl
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

use Getopt::Std;
use qpid_proton;
use async;

$Getopt::Std::STANDARD_HELP_VERSION = 1;

sub VERSION_MESSAGE() {}

sub HELP_MESSAGE() {
    print "Usage: send_async.pl [OPTIONS] <msg_0> <msg_1> ...\n";
    print "Options:\n";
    print "\t-a     - the message address (def. amqp://0.0.0.0)\n";
    print "\t-r     - the reply-to address: //<domain>[/<name>]\n";
    print "\t msg_# - a text string to send\n";
}

my %optons = ();
getopts("a:r:", \%options) or usage();

our $address = $options{a} || "amqp://0.0.0.0";
our $replyto = $options{r} || "~/#";

package async::Sender;

@ISA = (async::CallbackAdapter);

sub on_start {
    my ($self) = @_;
    my $message = $self->{_message};
    my $messenger = $self->{_messenger};
    my $args = $_[1] || ("Hello world!");

    print "Started\n";

    $message->clear;
    $message->set_address("amqp://0.0.0.0");
    $message->set_reply_to($replyto) if (defined($replyto));

    foreach $arg ($args) {
        $message->set_body($arg);
        if ($replyto) {
            $message->set_reply_to($replyto);
        }
        $self->send($message, "on_status");
    }

    $messenger->receive() if (defined($replyto));
}

sub on_status {
    my ($self) = @_;
    my $messenger = $self->{_messenger};
    my $status = $_[1] || "";

    print "Status: ", $status, "\n";
}

sub on_receive {
    my ($self) = @_;
    my $message = $_[1];
    my $text = $message->get_body || "[empty]";

    print "Received: " . $text . "\n";

    $self->stop();
}

sub on_stop {
    print "Stopped\n";
}


package main;

our $msgr = new qpid::proton::Messenger();
our $app = async::Sender->new($msgr);

$app->run;

File: examples/python/messenger/send_async.py
#!/usr/bin/python
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
from __future__ import print_function
import sys, optparse
from async import *

parser = optparse.OptionParser(usage="usage: %prog [options] <msg_1> ... <msg_n>",
                               description="simple message sender")
parser.add_option("-a", "--address", default="amqp://0.0.0.0",
                  help="address: //<domain>[/<name>] (default %default)")
parser.add_option("-r", "--reply_to", help="reply_to: //<domain>[/<name>]")

opts, args = parser.parse_args()
if not args:
  args = ["Hello World!"]

class App(CallbackAdapter):

    def on_start(self):
        print("Started")
        self.message.clear()
        self.message.address = opts.address
        self.message.reply_to = opts.reply_to
        for a in args:
            self.message.body = a
            self.send(self.message, self.on_status)

        if opts.reply_to:
            self.messenger.recv()

    def on_status(self, status):
        print("Status:", status)
        if not opts.reply_to or opts.reply_to[0] != "~":
            args.pop(0)
            if not args: self.stop()

    def on_recv(self, msg):
        print("Received:", msg)
        if opts.reply_to and opts.reply_to[0] == "~":
            args.pop(0)
            if not args: self.stop()

    def on_stop(self):
        print("Stopped")

a = App(Messenger())
a.run()
